### PR TITLE
mon, osd: set subsystem prefix for logging MonCap and OSDCap

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Set subsys and prefix for logging in MonCap and OSDCap

Figured out this set of changes while working on PR #41779.

With this patch following is how log entries from MonCap.cc look -

```
2023-04-10T22:15:25.746+0530 7fa5debf46c0 20 MonCap is_capable service=mon command= read addr v2:192.168.29.90:40719/0 on cap allow *
2023-04-10T22:15:25.746+0530 7fa5debf46c0 20 MonCap  allow so far , doing grant allow *
2023-04-10T22:15:25.746+0530 7fa5debf46c0 20 MonCap  allow all
```

And this is how same log entries from MonCap.cc look without this patch -

```
2023-04-10T22:23:32.214+0530 7f1fb27f46c0 20 is_capable service=mon command= read addr v2:192.168.29.90:40164/0 on cap allow *
2023-04-10T22:23:32.214+0530 7f1fb27f46c0 20  allow so far , doing grant allow *
2023-04-10T22:23:32.214+0530 7f1fb27f46c0 20  allow all
```

Here are the links to these log entries in the codebase -
https://github.com/ceph/ceph/blob/main/src/mon/MonCap.cc#L459
https://github.com/ceph/ceph/blob/main/src/mon/MonCap.cc#L471
https://github.com/ceph/ceph/blob/main/src/mon/MonCap.cc#L484

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mds/MDSAuthCaps.cc [hunk 1]
```diff
@@ -349,12 +349,13 @@ bool MDSAuthCaps::parse(CephContext *c, std::string_view str, ostream *err)
     return true;
   }
 
+  if (c != nullptr)
```
Comment (rzarzynski) [inline src/mds/MDSAuthCaps.cc:352]: Why we started setting the member conditionally? This excludes the ability to nullify `cct` (abstracting whether this feature is needed or not).

File: src/osd/OSDCap.h [hunk 2]
```diff
@@ -222,14 +222,15 @@ ostream& operator<<(ostream& out, const OSDCapGrant& g);
 
 
 struct OSDCap {
+  CephContext* cct = nullptr;
```
Comment (rzarzynski) [inline src/osd/OSDCap.h:225]: OK, so we slightly inflate (by one pointer) the size of `OSDCap` and `MonCap`. I think it's worth doing that but I assumed neither of types has tons of instances.

File: src/mds/MDSAuthCaps.cc (hunk 1 above)
Comment (rishabh-d-dave) [reply src/mds/MDSAuthCaps.cc:352]: `cct` is initialized to `nullptr` in `MDSAuthCaps.h`. So I was skipping assignment. But, yeah, it's unnecessary.

**Comment by rishabh-d-dave** [comment]
jenkins test make check arm64

**Review by batrick — CHANGES_REQUESTED** [review]
I support adding debugging facilities here, good initiative.

Am I correct that every caller of these caps is using `g_ceph_context`? I'd prefer to do away with this argument entirely and hard-code `g_ceph_context` wherever needed.

**Comment by rishabh-d-dave** [comment]
> I support adding debugging facilities here, good initiative.
> 
> Am I correct that every caller of these caps is using `g_ceph_context`? 

Yes. 

> I'd prefer to do away with this argument entirely and hard-code `g_ceph_context` wherever needed.

Hardcode as in create data member in class/struct, right?

I like this idea, I think its indeed better. @rzarzynski Any opinion?

**Comment by batrick** [comment]
> > I'd prefer to do away with this argument entirely and hard-code `g_ceph_context` wherever needed.
> 
> Hardcode as in create data member in class/struct, right?

No. `g_ceph_context` is an external variable global to all Ceph code. Just reference that.

**Comment by rzarzynski** [comment]
:+1: for `g_ceph_context`.

**Comment by rishabh-d-dave** [comment]
@batrick @rzarzynski Please take a look.

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/112181/consoleFull#440967684e4bfde06-44c1-4b3c-8379-d9ee175fb257

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/112318/

File: src/mds/MDSAuthCaps.h [hunk 3]
```diff
@@ -226,7 +225,7 @@ class MDSAuthCaps
 
   friend std::ostream &operator<<(std::ostream &out, const MDSAuthCaps &cap);
 private:
-  CephContext *cct = nullptr;
+  CephContext *cct = g_ceph_context;
```
Comment (batrick) [inline src/mds/MDSAuthCaps.h:228]: If we're going to use `g_ceph_context` (a global variable), why bother storing it in the cap structure?

**Review by batrick — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/112322/

**Review by batrick — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
@batrick Would it be worth to to make same changes (deleting `CephContext` instances and instead using `g_ceph_context`) over rest of the repo? There are lots of such instances. 
```
grep -rnI CephContext src/mon/ | wc -l
42
grep -rnI CephContext src/osd/ | wc -l
177
grep -rnI CephContext src/mds/ | wc -l
12
grep -rnI CephContext src/auth/ | wc -l
84
grep -rnI CephContext src/ | wc -l
3730
```

**Comment by rishabh-d-dave** [comment]
@batrick @rzarzynski 

make check is failing with following error which does not make sense because g_ceph_context is globally available. Also, locally the PR branch builds successfully. I also messaged @batrick on IRC last Friday few times about this.
```
src/crimson/CMakeFiles/crimson-common.dir/__/mon/MonCap.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/MonCap.cc
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/MonCap.cc:461:7: error: use of undeclared identifier 'g_ceph_context'
  if (g_ceph_context)
      ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/MonCap.cc:473:9: error: use of undeclared identifier 'g_ceph_context'
    if (g_ceph_context)
        ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/MonCap.cc:486:11: error: use of undeclared identifier 'g_ceph_context'
      if (g_ceph_context)
          ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/mon/MonCap.cc:496:11: error: use of undeclared identifier 'g_ceph_context'
      if (g_ceph_context)
          ^
4 errors generated.

```

**Comment by batrick** [comment]
Have you tried including the relevant header file?

**Comment by rishabh-d-dave** [comment]
@batrick 
> Have you tried including the relevant header file?

I'll try that. But it's strange that locally the build is successful but that's not the case with make check job.

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/112420/

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/112422/

**Comment by rishabh-d-dave** [comment]
Hi @gregsfortytwo 

Thanks for the reply. Although make check is running fine now I haven't demystified this. I am really stuck at this.

> > @batrick
> > > Have you tried including the relevant header file?
> > 
> > 
> > I'll try that. But it's strange that locally the build is successful but that's not the case with make check job.
> 
> You're probably not building Crimson, which is where those errors were showing up. And actually this error exactly here may mean that MonCap can't use g_ceph_context directly — I'm not sure that lives in crimson. @athanatos is that against the rules?

To build crimson locally do I need some special instructions? I usually pass lots of options to `do_cmake.sh` (to disable RBD, RGW, etc.). To reproduce this error, I also tried building this PR branch with no options to `do_cmake.sh`. Just `sudo install-deps.sh && ./do_cmake.sh && cd build && ninja -j 7` and yet I couldn't reproduce this error.

**Comment by rishabh-d-dave** [comment]
make check passed last time. Noting down the link to it. https://jenkins.ceph.com/job/ceph-pull-requests/112439/.

**Comment by athanatos** [comment]
./do_cmake.sh -DWITH_SEASTAR=ON to build crimson.  Anything used in crimson indeed cannot use g_ceph_context, so please leave it out of MonCap.

**Comment by rishabh-d-dave** [comment]
@athanatos 

> ./do_cmake.sh -DWITH_SEASTAR=ON to build crimson. Anything used in crimson indeed cannot use g_ceph_context, so please leave it out of MonCap.

Okay. I'll revert the changes in MonCap in this case.

On running `./do_cmake.sh -DWITH_SEASTAR=ON` I see the following error. Is anything more required? 

```
-- Performing Test StdAtomic_EXPLICIT_LINK                                                                                                                                                                                                    
-- Performing Test StdAtomic_EXPLICIT_LINK - Success                                                                                                                                                                                          
-- Found StdAtomic: -latomic                                                                                                                                                                                                                  
-- Found lksctp-tools: /usr/lib64/libsctp.so                                                                                                                                                                                                  
-- Found numactl: /usr/lib64/libnuma.so                                                                                                                                                                                                       
-- Checking for one of the modules 'yaml-cpp'                                                                                                                                                                                                 
-- Found yaml-cpp: /usr/lib64/libyaml-cpp.so (found suitable version "0.6.3", minimum required is "0.5.1")                                                                                                                                    
-- Found ragel: /usr/bin/ragel (found suitable version "7.0.0.12", minimum required is "6.10")                                                                                                                                                
-- Found PthreadSetName: 1                                                                                                                                                                                                                    
-- Found Valgrind: /usr/include                                                                                                                                                                                                               
CMake Error at src/seastar/CMakeLists.txt:782 (message):                                                                                                                                                                                      
  Sanitizers not found!                                                                                                                                                                                                                       
                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                              
-- Configuring incomplete, errors occurred!                                           
```

**Comment by athanatos** [comment]
You are most likely missing a package in your build environment -- one of the ones related to address sanitizer most likely.  You should also be using a recent version of gcc -- 12 if you can.

**Comment by ljflores** [comment]
This is marked as ready for testing (approved + "needs-qa"), but I'm seeing some comments that it might not be ready. Let me know if it's ready.

File: src/mds/MDSAuthCaps.h (hunk 3 above)
Comment (vshankar) [reply src/mds/MDSAuthCaps.h]: This comment has been taken care of.

**Comment by vshankar** [comment]
MDS changes LGTM.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
@ljflores this is ready now.

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests-arm64/39867/

**Review by vshankar — APPROVED** [review]

**Comment by ljflores** [comment]
Rados suite results: http://pulpito.front.sepia.ceph.com/?branch=wip-yuri7-testing-2023-04-19-1343

Failures, unrelated:
    1. https://tracker.ceph.com/issues/57755
    2. https://tracker.ceph.com/issues/58946
    3. https://tracker.ceph.com/issues/49888
    4. https://tracker.ceph.com/issues/59380
    5. https://tracker.ceph.com/issues/57754
    6. https://tracker.ceph.com/issues/55347
    7. https://tracker.ceph.com/issues/49287

Details:
    1. task/test_orch_cli: test_cephfs_mirror times out - Ceph - Orchestrator
    2. cephadm: KeyError: 'osdspec_affinity' - Ceph - Mgr - Dashboard
    3. rados/singleton: radosbench.py: teuthology.exceptions.MaxWhileTries: reached maximum tries (3650) after waiting for 21900 seconds - Ceph - RADOS
    4. rados/singleton-nomsgr: test failing from "Health check failed: 1 full osd(s) (OSD_FULL)" and "Health check failed: 1 filesystem is offline (MDS_ALL_DOWN)" - Ceph - RGW
    5. test_envlibrados_for_rocksdb.sh: update-alternatives: error: alternative path /usr/bin/gcc-11 doesn't exist - Ceph - RADOS
    6. SELinux Denials during cephadm/workunits/test_cephadm - Ceph - Orchestrator
    7. podman: setting cgroup config for procHooks process caused: Unit libpod-$hash.scope not found - Ceph - Orchestrator

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test make check arm64

**Comment by kotreshhr** [comment]
jenkins test make check arm64

**Review by rzarzynski — APPROVED** [review]
:+1:

**Pull request merged**